Include/String/String.h

```diff
@@ -53,51 +56,46 @@ class String {
         String toUpper() const;
         String toLower() const;
         String trim() const;
-        std::vector<String> split(const String& needle, bool removeEmptyEntries) const;
+        std::vector<String> split(const String& needleStr, bool removeEmptyEntries) const;
         static String join(const std::vector<String>& vect, const String& separator);
 
         std::cmatch regexMatch(const std::regex& pattern) const;
 
-        String unHex() const;
+        //String unHex() const;
 
-        String& operator=(const String& other);
-        String& operator+=(const String& other);
-
-        long long getHashCode() const;
+        uint64_t getHashCode() const;
         bool equals(const String& other) const;
         bool equalsIgnoreCase(const String& other) const;
         bool isEmpty() const;
 
     private:
         String(int size);
+        String(const String& other, int from, int cnt);
 
-        long long hashCode;
         int cCapacity = 0;
-        int strByteLength;
-        int strLength;
 
-        const static int shortStrCapacity = 16;
+        // Lazily evaluated.
+        mutable bool _hashCodeEvaluted;
+        mutable uint64_t _hashCode;
+        mutable int _strByteLength;
+        mutable int _strLength;
+
+        constexpr static int shortStrCapacity = 16;
 
         union _StringData {
             char shortStr[shortStrCapacity];
             char* longStr;
-            _StringData();
         } data;
 
-        void recalculateHashAndLength();
         void wCharToUtf8Str(const wchar* wbuffer);
-        static wchar utf8ToWChar(const char* cbuffer);
         void reallocate(int size);
-        void reset();
         char* cstrNoConst();
     };
 
     bool operator==(const String& a, const String& b);
     bool operator!=(const String& a, const String& b);
     const String operator+(const String& a, const String& b);
```

Comment:
Why is a const String returned here?

Reply:
To prevent the following from being valid:
`str1+str2 = "thing";`